Refactor Span SpanBuilder and SpanContext to use GRPC entities

danielamiao opened the pull request:
First piece of refactoring, changed all the span related classes to use GRPC:
- Refactored thrift entities into equivalent GRPC ones if exists
- Removed traceJoinId (no such thing in protobuf IDL)
- Added type to value of KeyValue (string vs. boolean)
 - For Number type, not sure if should extract int or double - would we lose precision here? Not exactly sure what the benefit is over just storing them as string values.
- Added reference entities to Span to denote childOf and followsFrom relationships since proto IDL defines them explicitly
- None of the corresponding unit tests have been refactored - will do them in subsequent PR

Refactoring to be expected in future PRs:
- AbstractTracer
- ClientMetrics
- Status
- All tests

danielamiao commented:
Also - CI won't build because not all refactoring is completed, just want to get your comments early on.

danielamiao commented:
Refactoring done for all classes, have not started factoring tests so CI still won't pass, but everything builds otherwise (probably doesn't actually work).

TODOs left:
- Unit tests are not refactored - not passing yet
- Need to clean up client class
- Need to catch exceptions more smartly

Pending questions for discussion:
- Separate KeyValue types for value? As mentioned, what is the value of doing this vs serializing everything as a string? Is it much slower?
- Add timing information to report? Currently, no duration is sent in the report - is it a problem? Doesn't look like it was ever sent even in Thrift.
- Add references/relationships to span? This was added since I saw the definition in the proto IDL but it does not exist in the thrift IDL, is this useful/important?

frenchfrywpepper commented on common/src/main/java/com/lightstep/tracer/shared/AbstractTracer.java:
```diff
@@ -259,9 +249,11 @@ private void disable() {
-                transport = null;
+            try {
+                client.shutdown();
+            } catch (InterruptedException e) {
```
It's not a great practice to swallow InterruptedExceptions (https://www.ibm.com/developerworks/library/j-jtp05236/)

danielamiao replied:
Thanks for catching this - I should probably handle the exception inside the client anyway.

frenchfrywpepper commented on common/src/main/java/com/lightstep/tracer/shared/AbstractTracer.java:
```diff
@@ -397,78 +389,63 @@ private boolean sendReportWorker(boolean explicitRequest) {
-            return false;
-        } catch (TException x) {
+        } catch (Exception e) {
+            // TODO Catch more specific grpc exceptions
```
+1 to the todo. Something more specific than Exception would be good.

bensigelman replied:
I am a strong +1 about catching more specific Exceptions. That said, we have learned from experience that it's also "good" to catch `Exception` as a final failsafe to protect the applications we are embedded within... customers really do not like it when lightstep client libs crash their process. :-/

danielamiao replied:
Yes, good point, I haven't gotten around to figuring out the specific exceptions thrown by GRPC client - will catch/handle those + a final Exception to avoid crashing customer application.

frenchfrywpepper commented:
Also, I think you're right about all that string serialization, doesn't seem like it's needed (at least not in this GRPC world.)

bensigelman commented on common/src/main/java/com/lightstep/tracer/shared/AbstractTracer.java:
```diff
@@ -99,28 +89,28 @@ public AbstractTracer(Options options) {
+        clockState = new ClockState();
+        clientMetrics = new ClientMetrics(Util.epochTimeMicrosToProtoTime(nowMicrosApproximate()));
+
+        auth = Auth.newBuilder();
```
nit:
```
auth = Auth.newBuilder().setAccessToken(options.accessToken)
```

(and similarly in other places where this pattern arises)

bensigelman commented on common/src/main/java/com/lightstep/tracer/shared/AbstractTracer.java:
```diff
@@ -53,23 +45,24 @@
     // shutdown).
     private final AtomicLong lastNewSpanMillis;
-    private ArrayList<SpanRecord> spans;
+    private ArrayList<Span> spans;
```
let's call this member `spanBuffer`... will maybe be clearer that this is only for serialized / finished spans that way.

danielamiao replied:
Oh, so the java classes are auto-generated from the proto file and Span is named so in that file. I don't think it's a good idea to change the name of an autogenerated file?

bensigelman commented on common/src/main/java/com/lightstep/tracer/shared/AbstractTracer.java:
```diff
@@ -477,7 +454,7 @@ private boolean sendReportWorker(boolean explicitRequest) {
 
             // TODO should probably prepend and do it in bulk
-            for (SpanRecord span : req.span_records) {
+            for (Span span : reqBuilder.getSpansList()) {
```
btw, this recovery path is debatable... it's also defensible to simply drop the data as long as we increment the client-dropped counter and report that information back to the collectors eventually.

bensigelman replied:
(to be clear, fine to leave it this way for now... I'm just trying to point out things where the Java client is maybe showing its age a bit)

bensigelman commented on common/src/main/java/com/lightstep/tracer/shared/CollectorClient.java:
```diff
@@ -0,0 +1,46 @@
+   * Constructor client for accessing CollectorService at {@code host:port}
+   */
+  public CollectorClient(String host, int port) {
+    this(ManagedChannelBuilder.forAddress(host, port).usePlaintext(true));
```
def don't want to default to plaintext when this gets merged into master, btw.

danielamiao replied:
👍

bensigelman commented on common/src/main/java/com/lightstep/tracer/shared/CollectorClient.java:
```diff
@@ -0,0 +1,46 @@
+import io.grpc.ManagedChannelBuilder;
+import java.util.concurrent.TimeUnit;
+
+public class CollectorClient {
```
I'm not clear to me that this abstraction layer is buying us much. It might make more sense if some of the state (counter logic, etc) from AbstractTracer moved in here, though.

danielamiao replied:
I have a TODO to investigate and properly implement this class later, I have a feeling the GRPC client will need to be more sophisticated (connection management?), warranting a separate class. For now I want to see a working GRPC connection to the collector first - if it really is *this* easy in the end I'll get rid of the class :)

bensigelman commented on common/src/main/java/com/lightstep/tracer/shared/Span.java:
```diff
@@ -84,11 +76,8 @@ public Span setTag(String key, Number value) {
-            } else {
-                record.addToAttributes(new KeyValue(key, value.toString()));
-            }
+            // TODO convert to number value? lose precision?
```
Yeah, we'll want to maintain the Number-ness of numeric tags as long as `KeyValue`'s value looks like a variadic type. Same with bools.

Ironically, everything is coerced to a string in the collector, but that's not the endgame.

danielamiao replied:
Alright, will do that then. I was originally more concerned about losing precision if say the number was a BigDecimal, using string will preserve the precision while using double won't. On second thought I can just convert Long and Double classes for now, and use string for all other number classes not explicitly defined in protobuf.

bensigelman commented on common/src/main/java/com/lightstep/tracer/shared/Util.java:
```diff
@@ -27,4 +28,15 @@ protected Random initialValue() {
+
+    static Timestamp epochTimeMicrosToProtoTime(long micros) {
+        Timestamp.Builder builder = Timestamp.newBuilder();
+        builder.setSeconds(micros / 1000 / 1000);
```
pedantic nit: can you either add some parens or just write `1000000` here? I got burned by an operator precedence bug once and told myself "never again" :)

bensigelman replied:
(same on the line below)

danielamiao commented:
Addressed PR comments, all unit tests now pass. Need to work on getting examples running/working.

TODOs:
- add to CollectorClient implementation
- catch more specific GRPC exceptions

danielamiao commented:
JRE example running, CI passes. 

TODO
Android examples

danielamiao commented:
Thanks, yup definitely will need some cleanup around exception handling before merging to master!
